README.md
[![](https://travis-ci.org/LUMC/fastqsplitter.svg?branch=develop)](
https://travis-ci.org/LUMC/fastqsplitter)

# fastqsplitter

A simple application to split FASTQ files.

Fastqsplitter reads a fastq
file. It then splits the reads over the designated output files.

This application does NOT work with multiline fastq sequences.

### Quickstart

install fastqsplitter:
```pip install fastqsplitter```

To split an input file `input_fastq.gz` into 3 different files.
`fastqsplitter -i input_fastq.gz 
-o split.1.fq.gz -o split.2.fq.gz -o split.3.fq.gz`

fastqsplitter uses the excellent [xopen library by @marcelm](
-https://github.com/marcelm/xopen). Therefore the input and output files 
compression is determined by the extension. Use ``.gz`` if output files should
be gzip compressed. 

### Documentation

More information on fastqsplitter can be found [on our readthedocs page](
https://fastqsplitter.readthedocs.io/).

CHANGELOG.rst
==========
Changelog
==========

.. Newest changes should be on top.

.. NOTE: This document is user facing. Please word the changes in such a way
.. that users understand how the changes affect the new version.

2.0.0-dev
-----------------
+ Redesigned CLI to make it much easier to use with streaming data.
+ Added an algorithm that can handle streaming data with no known input size.
+ Improved speed of the python algorithm. It is now 5 times faster than the
  old python algorithm. It is also 3 times faster than the cython algorithm
  from v1.2.0.
+ The cython parts of the code have been deprecated for easier installation
  and better platform compatibility.

1.2.0
-----------------
+ Enable pure python fallback so package can be installed on all systems.
+ Updated the documentation to reflect changes in speed because of the upstream
  improvements and the cythonizing of the algorithm in 1.1.0.
+ Upstream contributions to `xopen <https://github.com/marcelm/xopen>`_ have
  made the reading of gzipped fastq files significantly faster. Newer
  versions of xopen are now added as a requirement.

1.1.0
-----------------
+ Enable the building of wheels for the project now that Cython extensions
  are used. Thanks to @marcelm  for providing a working build script on
  https://github.com/marcelm/dnaio.
+ Cythonize the splitting algorithm. This reduces the overhead of the application
  up to 50% over the fastest native python implementation. Overhead is all the
  allocated cpu time that is not system time.

  This means splitting of uncompressed fastqs will be noticably faster
  (30% faster was achieved during testing). When splitting compressed
  fastq files into compressed split fastq files this change will not be much faster
  since all the gzip process will be run in a separate thread. Still when splitting
  a 2.3 gb gzipped fastq file into 3 gzipped split fastq files the speedup from
  the fastest python implementation was 14% in total cpu seconds. (Due to the
  multithreaded nature of the application wall clock time was reduced by only 3%).

1.0.0
-------------
+ Added documentation for fastqsplitter and set up readthedocs page.
+ Added tests for fastqsplitter.
+ Upstream contributions to xopen have improved fastqsplitter speed.
+ Initiated fastqsplitter.

tests/test_fastqsplitter.py
# Copyright (c) 2019 Leiden University Medical Center
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

import os
import sys
import tempfile
from pathlib import Path
from typing import Union

from Bio.SeqIO.QualityIO import FastqPhredIterator

from fastqsplitter import fastqsplitter, human_readable_to_int, main, \
    split_fastqs_round_robin, split_fastqs_sequentially

import pytest

import xopen

TEST_FILE = str(Path(__file__).parent / "data" / "test.fq.gz")
TEST_FILE_INVALID = str(Path(__file__).parent / "data" / "test_invalid.fq.gz")


def validate_fastq_gz(fastq: Union[str, Path]) -> int:
    """This function uses the biopython parser to validate a fastq file
    and outputs the number of fastqrecords."""

    number_of_records = 0
    with xopen.xopen(fastq, 'rt') as fastq_handle:
        fastq_iterator = FastqPhredIterator(fastq_handle)
        for _ in fastq_iterator:
            number_of_records += 1
    return number_of_records


# This also makes sure we have a valid test file.
RECORDS_IN_TEST_FILE = validate_fastq_gz(TEST_FILE)
with xopen.xopen(TEST_FILE, "rb") as fastq_handle:
    BYTES_IN_TEST_FILE = len(fastq_handle.read())


def test_invalid_test_file():
    # We need to make sure our test function indeed fails when a fastq file is
    # invalid.
    with pytest.raises(ValueError) as error:
        validate_fastq_gz(TEST_FILE_INVALID)
    error.match("Lengths of sequence and quality values differs")


@pytest.mark.parametrize(["input", "expected"],
                         [("1252", 1252),
                          ("64K", 64*1024),
                          ("64M", 64*1024*1024),
                          ("64G", 64*1024*1024*1024)]
                         )
def test_human_readable_to_int(input: str, expected: int):
    assert human_readable_to_int(input) == expected


def test_split_fastq_buffer_size_to_low():
    output_files = [str(tempfile.mkstemp(suffix=".fq")[1])
                    for _ in range(3)]
    with pytest.raises(ValueError) as error:
        split_fastqs_round_robin(TEST_FILE, output_files, buffer_size=512)
    error.match("at least 1024")


def test_split_fastq_no_output_files():
    with pytest.raises(ValueError) as error:
        split_fastqs_round_robin(TEST_FILE, [], buffer_size=512)
    error.match("at least 1")


@pytest.mark.parametrize("number_of_splits", list(range(1, 6)))
def test_split_fastqs_round_robin(number_of_splits: int):
    output_files = [str(tempfile.mkstemp(suffix=".fq")[1])
                    for _ in range(number_of_splits)]
    # Use a small buffer size for testing a small file.
    split_fastqs_round_robin(TEST_FILE, output_files, buffer_size=1024)

    expected_number_of_records = RECORDS_IN_TEST_FILE // number_of_splits

    total_records = 0
    for output_file in output_files:
        actual_records = validate_fastq_gz(output_file)
        total_records += actual_records
        # Check if distribution is even across output files.
        percentage_size = actual_records / expected_number_of_records
        assert percentage_size >= 0.98
        assert percentage_size <= 1.02
    assert total_records == RECORDS_IN_TEST_FILE


@pytest.mark.parametrize("max_size", [16*1024, 32*1024, 64*1024, 128*1024])
def test_split_fastqs_sequentially(max_size: int):
    prefix = tempfile.mktemp()
    buffer_size = 1024
    # The buffer = 1024. The target size is max_size - buffer size, and the
    # record is completed after that.
    # Record size is approx 256 characters per record.
    # So we expect 256 x 1.5 = 384 characters to be read on our record
    # completion on average.
    expected_length = BYTES_IN_TEST_FILE // (max_size - buffer_size + 384) + 1
    split_files = split_fastqs_sequentially(TEST_FILE,
                                            max_size=max_size,
                                            prefix=prefix,
                                            suffix=".fastq",
                                            buffer_size=buffer_size)
    assert len(split_files) == expected_length
    for file in split_files:
        validate_fastq_gz(file)
        assert os.stat(file).st_size <= max_size
        os.remove(file)  # cleanup


def test_fastqsplitter_number_of_files():
    prefix = tempfile.mktemp()
    number = 4
    output_files = fastqsplitter(TEST_FILE, prefix=prefix, number=number,
                                 buffer_size=1024, suffix=".fastq")
    expected_size = BYTES_IN_TEST_FILE / number
    assert len(output_files) == number
    for output_file in output_files:
        validate_fastq_gz(output_file)
        size_percentage = os.stat(output_file).st_size / expected_size
        assert size_percentage >= 0.98
        assert size_percentage <= 1.02


def test_fastqsplitter_output_files():
    number = 4
    output_files = [str(tempfile.mkstemp(suffix=".fq")[1])
                    for _ in range(number)]

    expected_size = BYTES_IN_TEST_FILE / number
    fastqsplitter(TEST_FILE, output=output_files, buffer_size=1024,
                  suffix=".fastq")
    for output_file in output_files:
        validate_fastq_gz(output_file)
        size_percentage = os.stat(output_file).st_size / expected_size
        assert size_percentage >= 0.98
        assert size_percentage <= 1.02
        os.remove(output_file)


def test_fastqsplitter_max_size_round_robin():
    prefix = tempfile.mktemp()
    max_size = 64 * 1024
    output_files = fastqsplitter(TEST_FILE, max_size=max_size, prefix=prefix,
                                 buffer_size=1024, round_robin=True)
    expected_number = os.stat(TEST_FILE).st_size // max_size + 1
    expected_size = os.stat(TEST_FILE).st_size // expected_number
    assert len(output_files) == expected_number
    for output_file in output_files:
        validate_fastq_gz(output_file)
        size_percentage = os.stat(output_file).st_size / expected_size
        assert size_percentage >= 0.98
        assert size_percentage <= 1.02
        os.remove(output_file)


def test_fastqsplitter_max_size_sequentially():
    buffer_size = 1024
    max_size = 64*1024
    min_size = max_size - buffer_size
    prefix = tempfile.mktemp()
    output_files = fastqsplitter(TEST_FILE, max_size=max_size, prefix=prefix,
                                 suffix=".fastq", buffer_size=buffer_size,
                                 round_robin=False)
    expected_number = BYTES_IN_TEST_FILE // max_size + 1
    assert len(output_files) == expected_number
    for output_file in output_files[:-1]:  # Last file will be different
        validate_fastq_gz(output_file)
        size = os.stat(output_file).st_size
        assert size >= min_size
        assert size <= max_size
        os.remove(output_file)
    validate_fastq_gz(output_files[-1])


def test_fastqsplitter_cannot_determine_output():
    with pytest.raises(ValueError) as error:
        fastqsplitter(TEST_FILE)
    error.match("Either")
    error.match("must be defined")


def test_fastqsplitter_error_sequential_no_max_size():
    with pytest.raises(ValueError) as error:
        fastqsplitter(TEST_FILE, round_robin=False)
    error.match("Maximum size")


def test_fastqsplitter_error_no_input_size():
    with pytest.raises(OSError) as error:
        fastqsplitter("/dev/urandom", max_size=64 * 1024)
    error.match("or empty input file")
    error.match("Cannot determine size")


def test_main(capsys):
    number_of_splits = 3
    prefix = tempfile.mktemp()
    output_files = [str(tempfile.mkstemp(suffix=".fq.gz")[1])
                    for _ in range(number_of_splits)]
    args = ["fastqsplitter", str(TEST_FILE), "-c", "5", "-t", "2", "-P",
            "-p", prefix]
    for output_file in output_files:
        args.append("-o")
        args.append(output_file)
    sys.argv = args[:]
    main()
    assert "\n".join(output_files) + "\n" == capsys.readouterr().out
    for output_file in output_files:
        validate_fastq_gz(output_file)
        os.remove(output_file)


def test_main_human_readable_conversion(capsys):
    prefix = tempfile.mktemp()
    sys.argv = ["fastqsplitter", str(TEST_FILE), "-m", "64K", "-P",
                "-p", prefix]
    main()
    output_files = capsys.readouterr().out.split()
    assert len(output_files) == 2
    for output_file in output_files:
        validate_fastq_gz(output_file)
        os.remove(output_file)


fastq_split.py
#!/usr/bin/env python3

import argparse
import contextlib
import gzip

from pathlib import Path
from typing import List

DEFAULT_COMPRESSION_LEVEL = 6


def argument_parser() -> argparse.ArgumentParser:
    parser = argparse.ArgumentParser()
    parser.add_argument("-i", "--input", type=Path, required=True)
    parser.add_argument("-o", "--output", action="append", type=Path,
                        required=True)
    parser.add_argument("-c", "--compression-level", type=int,
                        default=DEFAULT_COMPRESSION_LEVEL)
    return parser


def split_fastqs(input_file: Path, output_files: List[Path],
                 compression_level: int = DEFAULT_COMPRESSION_LEVEL,
                 group_size: int = 100):
    # contextlib.Exitstack allows us to open multiple files at once which
    # are automatically closed on error.
    # https://stackoverflow.com/questions/19412376/open-a-list-of-files-using-with-as-context-manager
    with contextlib.ExitStack() as stack:
        input_fastq = stack.enter_context(
            gzip.open(str(input_file), mode='rb'))  # type: gzip.GzipFile
        output_handles = [
                stack.enter_context(gzip.open(
                    filename=str(output_file),
                    mode='wb',
                    compresslevel=compression_level
                )) for output_file in output_files
        ]  # type: List[gzip.GzipFile]

        i = 0
        number_of_output_files = len(output_handles)
        fastq_empty = False
        while not fastq_empty:
            # By using modulo we pick one file at a time and make sure all
            # files are written to equally.
            file_to_write = output_handles[i % number_of_output_files]

            # Make sure a multiple of 4 lines is written as each record is
            # 4 lines.
            for j in range(group_size * 4):
                try:
                    # next(handle) equals handle.readline() but throws a
                    # StopIteration error if the lines is ''.
                    file_to_write.write(next(input_fastq))
                except StopIteration:
                    fastq_empty = True
                    break
            i += 1


def main():
    parser = argument_parser()
    parsed_args = parser.parse_args()
    split_fastqs(parsed_args.input,
                 parsed_args.output,
                 compression_level=parsed_args.compression_level)


if __name__ == "__main__":
    main()


build-wheels.sh
#!/bin/bash
#
# This file was adapted from https://github.com/marcelm/dnaio/blob/master/buildwheels.sh
# All credits to marcelm
#
# Build manylinux1 wheels. Based on the example at
# <https://github.com/pypa/python-manylinux-demo>
#
# For interactive tests:
#   docker run -it -v $(pwd):/io quay.io/pypa/manylinux1_x86_64 /bin/bash

set -xeuo pipefail

# For convenience, if this script is called from outside of a docker container,
# it starts a container and runs itself inside of it.
if ! grep -q docker /proc/1/cgroup; then
  # We are not inside a container
  docker pull quay.io/pypa/manylinux1_x86_64
  exec docker run --rm -v $(pwd):/io quay.io/pypa/manylinux2010_x86_64 /io/$0
fi

# Strip binaries (copied from multibuild)
STRIP_FLAGS=${STRIP_FLAGS:-"-Wl,-strip-all"}
export CFLAGS="${CFLAGS:-$STRIP_FLAGS}"
export CXXFLAGS="${CXXFLAGS:-$STRIP_FLAGS}"


# We don’t support Python 2.7 and 3.4
rm /opt/python/cp27*
rm /opt/python/cp34*

PYBINS="/opt/python/*/bin"
HAS_CYTHON=0
for PYBIN in ${PYBINS}; do
    ${PYBIN}/pip install Cython
#    ${PYBIN}/pip install -r /io/requirements.txt
    ${PYBIN}/pip wheel /io/ -w wheelhouse/
done

# Bundle external shared libraries into the wheels
for whl in wheelhouse/fastqsplitter-*.whl; do
    auditwheel repair "$whl" -w repaired/
done

# Created files are owned by root, so fix permissions.
mkdir -p /io/dist
chown -R --reference=/io/setup.py repaired/ /io/dist
mv repaired/*.whl /io/dist/


src/fastqsplitter/__init__.py
#!/usr/bin/env python3

"""
This fastqsplitter module provides a program to split fastq files. It
implements two algorithms to do so, one round-robin and one sequential
algorith.
"""

# Copyright (c) 2019 Leiden University Medical Center
#
# Permission is hereby granted, free of charge, to any person obtaining a copy
# of this software and associated documentation files (the "Software"), to deal
# in the Software without restriction, including without limitation the rights
# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
# copies of the Software, and to permit persons to whom the Software is
# furnished to do so, subject to the following conditions:
#
# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.
#
# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
# SOFTWARE.

import argparse
import contextlib
import io
import os
import sys
from typing import List, Optional

# xopen opens files as normal files, gzip files, bzip2 files or xz files
# depending on extension.
import xopen

try:
    import fcntl
    with open("/proc/sys/fs/pipe-max-size", "rt") as handle:
        MAX_PIPE_SIZE: Optional[int] = int(handle.read())
    # Python 3.10 will include F_SETPIPE_SZ parameter
    # https://github.com/python/cpython/pull/21921
    # Linux F_SETPIPE_SZ = 1031.
    if not hasattr(fcntl, "F_SETPIPE_SZ") and sys.platform == "linux":
        setattr(fcntl, "F_SETPIPE_SZ", 1031)
except (ImportError, FileNotFoundError):
    fcntl = None  # type: ignore
    MAX_PIPE_SIZE = None  # type: ignore

# Choose 1 as default compression level. Speed is more important than filesize
# in this application.
DEFAULT_COMPRESSION_LEVEL = 1
# 48K seems a good default. This works both for compressed reads and
# uncompressed reads. See docs/benchmark_results.txt.
DEFAULT_BUFFER_SIZE = 48 * 1024
# With one thread per file (pigz -p 1) pigz uses way less virtual memory
# (10 vs 300 MB for 4 threads) and total CPU time is also decreased.
# Example: 2.3 gb input file file. Five output files.
# TT=Total cpu time. RT=realtime
# Compression level 1: threads=1 RT=58s TT=3m33, threads=4 RT=57s TT=3m45
# Compression level 5: threads=1 RT=1m48 TT=7m53, threads=4 RT=1m23, TT=8m39
# But this is on a 8 thread machine. When using less threads RT will go towards
# TT. Default compression is 1. So default threads 1 makes the most sense.
DEFAULT_THREADS_PER_FILE = 1
DEFAULT_SUFFIX = ".fastq.gz"
STDIN = "/dev/stdin" if os.name == "posix" else None
SIZE_SUFFIXES = {"K": 1024 ** 1, "M": 1024 ** 2, "G": 1024 ** 3}


def optimize_xopen_handle(file_handle):
    if (isinstance(file_handle, xopen.PipedGzipReader) or
            isinstance(file_handle, xopen.PipedGzipWriter)):
        if hasattr(fcntl, "F_SETPIPE_SZ") and MAX_PIPE_SIZE:
            # Set pipe size to max value
            fcntl.fcntl(file_handle._file.fileno(), fcntl.F_SETPIPE_SZ,
                        MAX_PIPE_SIZE)
        return file_handle._file
    return file_handle


def argument_parser() -> argparse.ArgumentParser:
    """Argument parser for the fastqsplitter application"""
    parser = argparse.ArgumentParser()
    parser.add_argument("input", type=str, default=STDIN, nargs="?",
                        help="The fastq file to be scattered.")
    parser.add_argument("-p", "--prefix", type=str,
                        help="The prefix for the output files.")
    parser.add_argument("-s", "--suffix", type=str, default=DEFAULT_SUFFIX,
                        help="The default suffix for the output files. The "
                             "extension determines which compression is used. "
                             "'.gz' for gzip, '.bz2' for bzip2, '.xz' for xz. "
                             "Other extensions will use no compression.")
    output_group = parser.add_mutually_exclusive_group(required=True)
    output_group.add_argument("-n", "--number", type=int,
                              help="Specify the number of output files which "
                                   "to split over. Fastq records will be "
                                   "distributed using a round-robin method.")
    output_group.add_argument(
        "-o", "--output", action="append", type=str,
        help="Scatter over these output files. Multiple -o flags can be used. "
             "The extensions determine which compression algorithm will be "
             "used. '.gz' for gzip, '.bz2' for bzip2, '.xz' for xz. Other "
             "extensions will use no compression. Fastq records will be "
             "distributed using a round-robin method.")
    output_group.add_argument(
        "-m", "--max-size", type=str,
        help="In round robin mode, determines the number of output files by "
             "dividing the input size "
             "by max size and distribute the fastq records in a round robin "
             "fashion over these files. WARNING: if compression differs "
             "between input and output files, this will not work properly. "
             "In sequential mode this is the maximum number of bytes written "
             "to each output file. NOTE: This is the size *before* "
             "compression (if applied). As a rule of thumb multiply by 0.38 "
             "to get the actual filesize when using gzip compression.")

    # What is a good one-letter symbol for --no-round-robin?
    parser.add_argument("-S", "--sequential", action="store_false",
                        dest="round_robin",
                        help="Do not use round-robin but create output files "
                             "sequentially instead. Default when using "
                             "stdin. Max size should be set.")
    parser.add_argument("-c", "--compression-level", type=int,
                        default=DEFAULT_COMPRESSION_LEVEL,
                        help="Only applicable when output files have a '.gz' "
                             "extension. Default={0}"
                        .format(DEFAULT_COMPRESSION_LEVEL))
    parser.add_argument("-t", "--threads-per-file", type=int,
                        default=DEFAULT_THREADS_PER_FILE,
                        help="Set the number of compression threads per output"
                             " file. NOTE: more threads are only useful when "
                             "using a compression level > 1. To use "
                             "fastqsplitter in single-threaded mode choose "
                             "0. Default={0}."
                             "".format(DEFAULT_THREADS_PER_FILE))
    parser.add_argument("-P", "--print", action="store_true",
                        help="Print output files to stdout for easier usage "
                             "in scripts.")

    # BELOW ARGUMENTS ARE FOR BENCHMARKING AND TESTING PURPOSES
    parser.add_argument("-b", "--buffer-size", type=str,
                        default=str(DEFAULT_BUFFER_SIZE),
                        help=argparse.SUPPRESS)
    return parser


def human_readable_to_int(number_string: str) -> int:
    """
    Convert a string such as '64K' or '128M' to an integer.
    64K == 64 * 1024. Integers without suffix are converted as is.
    """
    for suffix, multiplier in SIZE_SUFFIXES.items():
        if number_string.endswith(suffix):
            return int(number_string.strip(suffix)) * multiplier
    return int(number_string)


def _read_until_new_fastq_record(input_handle: io.BufferedReader) -> bytes:
    """
    Reads the input handle until the start of a fastq record or has
    reached the EOF. This ensures the input handle is at the position of a new
    record. Returns the bytes read until the position.
    :param input_handle: an io.BufferedReader or similar object.
    :return: All bytes that have been read up to the start of the new fastq
    record
    """
    missing_record_lines = []  # type: List[bytes]
    while True:
        line = input_handle.readline()
        if line == b"":  # EOF
            return b"".join(missing_record_lines)
        missing_record_lines.append(line)
        # Both @ and + can occur in the quality line while they also indicate
        # the start of the record and the middle respectively.
        # Checks below ensure at least one complete record is returned.
        if line.startswith(b"@"):
            sequence_line = input_handle.readline()
            missing_record_lines.append(sequence_line)
            if sequence_line.strip().isalpha():  # Valid sequence line
                plus_line = input_handle.readline()
                missing_record_lines.append(plus_line)
                if plus_line.startswith(b"+"):
                    quality_line = input_handle.readline()
                    missing_record_lines.append(quality_line)
                    lengths_match = len(quality_line) == len(sequence_line)
                    # Peek returns a variable number of characters (but at
                    # least 1 in the case below). So we check with startswith.
                    new_record_start = input_handle.peek(1).startswith(b"@")
                    if lengths_match and new_record_start:
                        return b"".join(missing_record_lines)


def split_fastqs_round_robin(
        input_file: str, output_files: List[str],
        compression_level: int = DEFAULT_COMPRESSION_LEVEL,
        buffer_size: int = DEFAULT_BUFFER_SIZE,
        threads_per_file: int = DEFAULT_THREADS_PER_FILE) -> None:
    """
    Split a fastq file over multiple output files in a round robin fashion.
    :param input_file: The file to be split.
    :param output_files: The files receiving the split parts
    :param compression_level: Which compression level to use if applicable
    :param buffer_size: The buffer size. The granularity at which fastq records
    are distributed.
    :param threads_per_file: How many threads xopen should use to open the
    file.
    """
    if len(output_files) < 1:
        raise ValueError("The number of output files should be at least 1.")
    if buffer_size < 1024:
        # This value is arbitrary, but really low values such as 5 or 30 don't
        # make sense.
        raise ValueError("The buffer size should be at least 1024.")

    # contextlib.Exitstack allows us to open multiple files at once which
    # are automatically closed on error.
    # https://stackoverflow.com/questions/19412376/open-a-list-of-files-using-with-as-context-manager
    with contextlib.ExitStack() as stack:
        input_handle = stack.enter_context(
            xopen.xopen(input_file, mode='rb', threads=threads_per_file))
        input_handle = optimize_xopen_handle(input_handle)
        output_handles = [stack.enter_context(xopen.xopen(
                filename=output_file,
                mode='wb',
                compresslevel=compression_level,
                threads=threads_per_file
            )) for output_file in output_files
        ]  # type: List[io.BufferedWriter]
        output_handles = [optimize_xopen_handle(output_handle)
                          for output_handle in output_handles]

        group_number = 0
        number_of_output_files = len(output_files)

        while True:
            read_buffer = input_handle.read(buffer_size)
            if read_buffer == b"":
                return

            # Read the input until the start of a new record.
            completed_record = _read_until_new_fastq_record(input_handle)
            output_handles[group_number].write(read_buffer + completed_record)
            # Set the group number for the next group to be written.
            group_number += 1
            # cycle back to the start when we have written the last file.
            if group_number == number_of_output_files:
                group_number = 0


def _sequential_splitter(input_handle: io.BufferedReader,
                         output_handle: io.BufferedReader,
                         max_size: int,
                         buffer_size: int = DEFAULT_BUFFER_SIZE) -> int:
    """
    Reads max_size bytes from an input_handle and writes it to output_handle
    reading buffer_size bytes at the time. Ensures a complete fastq record
    is at the end of each file.
    :return: The number of bytes written.
    """
    target_size = max_size - buffer_size
    total_size = 0
    while True:
        read_buffer = input_handle.read(buffer_size)
        if read_buffer == b"":
            return total_size
        output_handle.write(read_buffer)
        total_size += buffer_size
        if total_size >= target_size:
            # Complete the record
            completed_record = _read_until_new_fastq_record(input_handle)
            output_handle.write(completed_record)
            return total_size + len(completed_record)


def split_fastqs_sequentially(
        input_file: str,
        max_size: int,
        prefix: str = "split.",
        suffix: str = DEFAULT_SUFFIX,
        buffer_size: int = DEFAULT_BUFFER_SIZE,
        compression_level: int = DEFAULT_COMPRESSION_LEVEL,
        threads_per_file: int = DEFAULT_THREADS_PER_FILE) -> List[str]:
    """
    Read an input file and create a new split output file for every
    max_size bytes read.
    :param input_file: The input fastq
    :param max_size: The maximum size of bytes that should be in the output
    files.
    :param prefix: Prefix for the output files.
    :param suffix: Suffix for the output files.
    :param buffer_size: How much data should be read at once.
    :param compression_level: The compression level if a '.gz' suffix is used.
    :param threads_per_file: The number of compressen threads per file.
    :return: A list of written files.
    """
    if max_size < buffer_size:
        raise ValueError("Maximum size {0} should be larger than buffer size "
                         "{1}.".format(max_size, buffer_size))

    with xopen.xopen(input_file, mode="rb", threads=threads_per_file
                     ) as input_fastq:
        group_number = 0
        written_files = []  # type: List[str]
        while True:
            if input_fastq.peek(0) == b"":  # Quit if there are no bytes left
                return written_files
            filename = prefix + str(group_number) + suffix
            group_number += 1  # Increase group_number for the next file
            with xopen.xopen(filename, mode="wb",
                             compresslevel=compression_level,
                             threads=threads_per_file) as output_fastq:
                _sequential_splitter(input_fastq, output_fastq,
                                     max_size,
                                     buffer_size=buffer_size)
                written_files.append(filename)


def fastqsplitter(input: str,
                  output: Optional[List[str]] = None,
                  number: Optional[int] = None,
                  max_size: Optional[int] = None,
                  prefix: Optional[str] = None,
                  suffix: str = DEFAULT_SUFFIX,
                  buffer_size: int = DEFAULT_BUFFER_SIZE,
                  compression_level: int = DEFAULT_COMPRESSION_LEVEL,
                  threads_per_file: int = DEFAULT_THREADS_PER_FILE,
                  round_robin: bool = True) -> List[str]:
    """
    Splits fastq files sequentially or round_robin depending on the given
    parameters. Creates files of the from <prefix><number><suffix>.
    :param input: The input fastq file.
    :param output: An optional list of output files if filenames should be
    determined before hand.
    :param number: Optional number of output files.
    :param max_size: Try to determine the number of output files by the size
    of the input file and the maximum size. Or if round_robin = false this is
    the maximum amount of bytes written.
    :param prefix: The prefix for the output files.
    :param suffix: The suffix for the output files. ".gz" files are gzip
    compressed, ".xz" xz compressed and ".bzip2" bzip2 compressed.
    :param buffer_size: The granularity with which the fastq files should be
    distributed.
    :param compression_level: The compression level to use, if applicable.
    :param threads_per_file: The amount of threads xopen should use to open
    the file.
    :param round_robin: If set to false will force the sequential method if
    a file is given.
    :return: The list of output files written.
    """
    default_prefix = os.path.basename(
        input).rstrip(".gz").rstrip(".fastq").rstrip(".fq") + "."
    prefix = prefix if prefix is not None else default_prefix

    if not round_robin or (input == STDIN and max_size is not None):
        if max_size is None:
            raise ValueError("Maximum size must be set when splitting files "
                             "sequentially (not using round-robin).")
        return split_fastqs_sequentially(
            input_file=input,
            max_size=max_size,
            prefix=prefix,
            suffix=suffix,
            buffer_size=buffer_size,
            compression_level=compression_level,
            threads_per_file=threads_per_file)

    if output:
        output_files = output
    else:
        if max_size is not None:
            input_size = os.stat(input).st_size
            if input_size == 0:
                raise OSError("Cannot determine size of input file or "
                              "empty input file: {0}.".format(input))
            number = input_size // max_size + 1
        elif not number:
            raise ValueError("Either a maximum size or a number of files or "
                             "a list of output files must be defined.")
        output_files = [prefix + str(i) + suffix for i in range(number)]

    split_fastqs_round_robin(input, output_files,
                             compression_level=compression_level,
                             threads_per_file=threads_per_file,
                             buffer_size=buffer_size)
    return output_files


def main():
    """Fastqsplitter program"""
    parser = argument_parser()
    # convert argparse.Namespace to dictionary
    kwargs = vars(parser.parse_args())
    max_size = kwargs.pop("max_size")
    if max_size is not None:
        max_size = human_readable_to_int(max_size)
    buffer_size = human_readable_to_int(kwargs.pop("buffer_size"))
    print_to_stdout = kwargs.pop("print")
    # kwargs correspond to fastqsplitter function inputs.
    output_files = fastqsplitter(max_size=max_size,
                                 buffer_size=buffer_size,
                                 **kwargs)
    if print_to_stdout:
        print("\n".join(output_files))


if __name__ == "__main__":  # pragma: no cover
    main()
